Don't fix Python patch version

Sjors opened the pull request:
`.python-version` always matches the minimum supported Python version.
It's main purpose is to catch accidental use of too modern syntax
in scripts and functional tests.

We (currently) don't specify a minimum patch version, so it's not
necessary to do so here. The minor verion is enough.

This also avoids requiring users to keep a potentially unsafe old
patch version installed.

The linter CI job used `python_build` (part of PyEnv) which requires specifying an exact Python version. The first commit changes it to use the higher level `pyenv install` which will pick the most recent patch version it knows about.

The original stated reason for using the lower level `python_build` https://github.com/bitcoin/bitcoin/pull/26716#issuecomment-1356314359 was that it allowed specifying `CC=clang` which needs fewer resources, but `pyenv install` can do that as well.  

See also https://github.com/bitcoin-core/HWI/pull/695

Sjors commented:
It's not really a problem for me, but I noticed HWI stopped using patch versions in https://github.com/bitcoin-core/HWI/pull/695 and I don't think there's a good reason for us to insist on one.

Everywhere else in the project we only specify the Python minor version. And if there's ever a security issue with the patch version we picked, we probably should go and update it (which we've never done).

fanquake commented:
> but then not just using pyenv install

`pyenv install` is a wrapper around `python-build`. Looks like it was used to use slightly less resources.

Sjors commented:
It's defined here: https://github.com/pyenv/pyenv/blob/master/plugins/python-build/bin/pyenv-install

Which internally calls `pyenv-latest --known`: https://github.com/pyenv/pyenv/blob/master/libexec/pyenv-latest

Which it turn calls `python-build --definitions` as I did.

fanquake commented:
The PR description and commit message will need re-writing, to explain the motivation, as the current motivaiton is just user error (updating python and not reinstalling dependencies).

Sjors committed:
ci: install Python using pyenv for linter

Using `pyenv install` avoids the need to specify a patch version,
which will be dropped in the next commit.

The original reason for using the lower level `python_build`, which
is part of PyEnv, was that it allowed specifying CC=clang, which
needs fewer resources. But `pyenv install` supports that too.

Sjors committed:
Don't pin Python patch version

.python-version always matches the minimum supported Python version.
It's main purpose is to catch accidental use of too modern syntax
in scripts and functional tests.

We (currently) don't specify a minimum patch version, so it's not
necessary to do so here. The minor verion is enough.

This also avoids requiring users to keep a potentially unsafe old
patch version installed.

Sjors commented:
Original discussion: https://github.com/bitcoin/bitcoin/pull/26716#issuecomment-1356314359

> I reckon the lower memory usage is due to using clang instead of gcc.

`pyenv install` also honors `CC=clang` (tested by setting `CC=fkfkkf` which fails)

> it'd be less lines of code.

It removes a few lines in the installers, but adds some for `pyenv global` and `pyenv init`.

Using PyEnv directly does make it easier to use multiple Python versions, in case any of our linters would benefit from a more recent version.

I updated the commit message.

maflcko commented:
Seems fine to change. I just recall another proposed change in https://github.com/bitcoin/bitcoin/pull/34391#discussion_r2732056569  which conflicts with this pull?

Again, seems fine. It would just not be ideal if this was changing code that is removed in the next commit/pull anyway.

maflcko commented:
Yes, it modifies the same lines of code, which is  a git merge conflict.
